aspnetcore/mvc/models/file-uploads.md

```diff
@@ -1,195 +1,656 @@
+
+> [!WARNING]
+> The validation processing methods demonstrated in the sample app don't scan the content of uploaded files. In most production scenarios, a virus/malware scanner API is used on the file before making the file available to users or other systems.
+
+### Content validation
+
+Use a third party virus/malware scanning API on uploaded content.
+
+Scanning files is demanding on server resources in high volume scenarios. If request processing performance is diminished due to file scanning, consider offloading the scanning work to a [background service](xref:fundamentals/host/hosted-services), possibly a service running on a server different from the app's server. Typically, uploaded files are held in a quarantined area until the background virus scanner checks them. When a file passes, the file is moved to the normal file storage location. These steps are usually performed in conjunction with a database record that indicates the scanning status of a file. By using such an approach, the app and app server remain focused on responding to requests.
+
+### File extension validation
+
+The uploaded file's extension should be checked against a list of permitted extensions. The sample app uses an array to determine if an extension is valid:
+
+```csharp
+private string[] permittedExtensions = { ".txt", ".pdf", };
+
+var ext = Path.GetExtension(uploadedFileName);
+
+if (string.IsNullOrEmpty(ext) || !permittedExtensions.Contains(ext))
+{
+    // The extension is invalid ... discontinue processing the file
+}
+```
+
+### File signature validation
 
-The following example demonstrates using JavaScript/Angular to stream to a controller action. The file's antiforgery token is generated using a custom filter attribute and passed in HTTP headers instead of in the request body. Because the action method processes the uploaded data directly, model binding is disabled by another filter. Within the action, the form's contents are read using a `MultipartReader`, which reads each individual `MultipartSection`, processing the file or storing the contents as appropriate. Once all sections have been read, the action performs its own model binding.
+A file's signature is determined by the first few bytes that make up the file. These bytes can be used to indicate if the extension matches the content of the file. The sample app checks file signatures for common file types. In the following example, the file signature for a Microsoft Word document is checked against the file:
 
-The initial action loads the form and saves an antiforgery token in a cookie (via the `GenerateAntiforgeryTokenCookieForAjax` attribute):
+```csharp
+private Dictionary<string, byte[]> _fileSignature = 
+    new Dictionary<string, byte[]>
+{
+    { ".docx", new byte[] { 0x50, 0x4B, 0x03, 0x04 } },
+};
+
+using (var reader = new BinaryReader(uploadedFileData))
```

Comment:
You could use a `StreamReader`, no need for `BinaryReader`. 

```C#
using var reader = new StreamReader();
var headerBytes = reader.ReadBytes(signatures.Values.Max(m => m.Length));

foreach (var (key, value) in signatures)
{
   if (headerBytes.Slice(0, value.Length).SequenceEquals(value))
   {
      // Success!
   }
}
```

Reply:
^ dont know if this compiles

Reply:
Can u `ReadBytes` on a `StreamReader`? IntelliSense is hating on it. Is that another C# 8.0 thing? Maybe I'm missing a namespace?

Your version defeats one of the goals. The current code makes sure that not only it matches *a signature in the dict* but that it matches the *one specific signature for that file extension*. I think that's important, but a dev would be welcome to go another way with it.

Reply:
```csharp
var signatures = _fileSignature[ext];
var headerBytes = reader.ReadBytes(signatures.Max(m => m.Length));

return signatures.Any(signature => 
    headerBytes.Take(signature.Length).ToArray().SequenceEqual(signature));
```

Reply:
You're right. You could use the raw stream to read it, not `StreamReader`, but that's a bit more involved.